devito/core/gpu_openmp.py

```diff
@@ -212,6 +220,82 @@ def _(iet):
         return iet, {}
 
 
+@iet_pass
+def initialize(iet, **kwargs):
+    """
+    Initialize the OpenMP environment.
+    """
+    devicenum = Symbol(name='devicenum')
+
+    @singledispatch
+    def _initialize(iet):
+        comm = None
+
+        for i in iet.parameters:
+            if isinstance(i, MPICommObject):
+                comm = i
+                break
+
+        if comm is not None:
+            rank = Symbol(name='rank')
+            rank_decl = LocalExpression(DummyEq(rank, 0))
+            rank_init = Call('MPI_Comm_rank', [comm, Byref(rank)])
+
+            ngpus = Symbol(name='ngpus')
+            call = Function('omp_get_num_devices')()
+            ngpus_init = LocalExpression(DummyEq(ngpus, call))
+
+            devicenum_init = LocalExpression(DummyEq(devicenum, rank % ngpus))
+
+            body = [rank_decl, rank_init, ngpus_init, devicenum_init]
+
+            init = List(header=c.Comment('Begin of OpenMP5.0+MPI setup'),
+                        body=body,
+                        footer=(c.Comment('End of OpenMP5.0+MPI setup'), c.Line()))
+        else:
+            devicenum_init = LocalExpression(DummyEq(devicenum, 0))
+            body = [devicenum_init]
+
+            init = List(header=c.Comment('Begin of OpenMP5.0 setup'),
+                        body=body,
+                        footer=(c.Comment('End of OpenMP5.0 setup'), c.Line()))
+
+        iet = iet._rebuild(body=(init,) + iet.body)
+
+        return iet
+
+    @_initialize.register(ElementalFunction)
+    @_initialize.register(MPICallable)
+    def _(iet):
+        return iet
+
+    iet = _initialize(iet)
+
+    return iet, {'args': devicenum}
+
+
+@iet_pass
+def mpi_gpu_direct(iet, **kwargs):
+    """
+    Modify MPI Callables to enable multiple GPUs performing GPU-Direct communication.
+    """
+    def _mpi_gpu_direct(iet):
+        mapper = {}
+        # Modify MPI_Isend and MPI_Irecv to perform GPU Direct communication
+        for node in FindNodes((IsendCall, IrecvCall)).visit(iet):
+            header = c.Pragma('omp target data use_device_ptr(%s) device(devicenum)' %
+                              node._args['parameters'][0].name)
```

Comment:
just `node.name` ?

see [here](https://github.com/devitocodes/devito/blob/master/devito/ir/iet/nodes.py#L238)

Reply:
It must be the name of the buffer rather than the name of the node:

(Pdb) node.name
'MPI_Irecv'
(Pdb) node._args['parameters'][0].name
'bufs'
[...]
(Pdb) node.name
'MPI_Isend'
(Pdb) node._args['parameters'][0].name
'bufg'

Reply:
oh yeah sorry. Then just `node.parameters[0].name` ?

Reply:
`AttributeError: 'IrecvCall' object has no attribute 'parameters'`
Are you happy with `node.arguments[0].name` instead?

Reply:
yes, sorry, I meant arguments oc

Reply:
I'll push that together with the next changes.